src/middlewared/middlewared/plugins/pool.py

```diff
@@ -4182,6 +4184,137 @@ def __handle_zfs_set_property_error(self, e, properties_definitions):
+            ['zfs', 'userspace', '-H', '-n', '-p', '-o', 'name,used,quota', id],
+            check=False
+        )
+        if userquota.returncode != 0:
+            raise CallError('Failed to get userquota for %s: [%s]',
+                            id, userquota.stderr.decode())
+
+        for quota in userquota.stdout.decode().splitlines():
+            entry = {
+                'uid': None,
+                'username': None,
+                'quota': 0,
+                'used': 0,
+            }
+            m = RE_USERSPACE.match(quota)
+            if m is None:
+                self.logger.debug('Invalid userspace quota: %s', quota)
+                continue
+
+            entry.update(m.groupdict())
+
+            try:
+                entry['username'] = (
+                    await self.middleware.call('user.get_user_obj', {'uid': entry['uid']})
+                )['pw_name']
+
+            except Exception:
+                self.logger.debug('Unable to resolve uid %d to name', entry['uid'])
+                pass
+
+            quota_list.append(entry)
+
+        return filter_list(quota_list, filters, options)
+
+    @accepts(
+        Str('id', required=True),
+        Ref('query-filters'),
+        Ref('query-options'),
+    )
+    @item_method
+    async def get_groupquota(self, id, filters, options):
+        """
+        Returns a list of group quotas on the ZFS dataset `id`. Supports
+        `query-filters` and `query-options`.
+        """
+        quota_list = []
+        groupquota = await run(
+            ['zfs', 'groupspace', '-H', '-n', '-p', '-o', 'name,used,quota', id],
+            check=False
+        )
+        if groupquota.returncode != 0:
+            raise CallError('Failed to get groupquota for %s: [%s]',
+                            id, groupquota.stderr.decode())
+
+        for quota in groupquota.stdout.decode().splitlines():
+            entry = {
+                'gid': None,
+                'groupname': None,
+                'quota': 0,
+                'used': 0,
+            }
+            m = RE_GROUPSPACE.match(quota)
+            if m is None:
+                self.logger.debug('Invalid groupspace quota: %s', quota)
+                continue
+
+            entry.update(m.groupdict())
+
+            try:
+                entry['groupname'] = (
+                    await self.middleware.call('group.get_group_obj', {'gid': entry['gid']})
+                )['gr_name']
+
+            except Exception:
+                self.logger.debug('Unable to resolve gid %d to name', entry['gid'])
+                pass
+
+            quota_list.append(entry)
+
+        return filter_list(quota_list, filters, options)
+
+    @accepts(
+        Str('id', required=True),
+        List('quotas', items=[
+            Dict(
+                'userspace_quota',
+                Str('type', enum=['USER', 'GROUP'], required=True),
+                Str('id', required=True),
+                Str('quota', required=True),
```

Comment:
We should use Int schema here which will validate that `quota` indeed is an integer